debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/cluster/MongoDbReplicaSetIT.java

```diff
@@ -0,0 +1,128 @@
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+import org.bson.BsonDocument;
+import org.bson.Document;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mongodb.ConnectionString;
+import com.mongodb.client.MongoChangeStreamCursor;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.changestream.ChangeStreamDocument;
+import com.mongodb.connection.ServerDescription;
+import com.mongodb.internal.selector.ReadPreferenceServerSelector;
+
+/**
+ * @see <a href="https://issues.redhat.com/browse/DBZ-5857">DBZ-5857</a>
+ */
+public class MongoDbReplicaSetIT {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbReplicaSetIT.class);
+
+    @Test
+    public void testCluster() throws InterruptedException {
+        try (var cluster = replicaSet().build()) {
+            LOGGER.info("Starting {}...", cluster);
+            cluster.start();
+
+            String readPreference = "primary";
+            var connectionString = new ConnectionString(cluster.getConnectionString() + "/?readPreference=" + readPreference);
+            LOGGER.info("Connecting to cluster: {}", connectionString);
+            try (var client = MongoClients.create(connectionString)) {
+                LOGGER.info("Connected to cluster: {}", client.getClusterDescription());
+                var collection = setup(client);
+
+                run(cluster, client, collection);
+            }
+        }
+    }
+
+    private void run(MongoDbReplicaSet cluster, MongoClient client, MongoCollection<Document> collection) throws InterruptedException {
+        try (var cursor = collection.watch().batchSize(1).cursor()) {
+            // Write 2 docs
+            collection.insertOne(Document.parse("{username: 'user" + 1 + "', name: 'User " + 1 + "'}"));
+            collection.insertOne(Document.parse("{username: 'user" + 2 + "', name: 'User " + 2 + "'}"));
+
+            // Read one
+            LOGGER.info("{}", cursor.next());
+
+            // Force a promotion that invalidates read preference on primary
+            LOGGER.info("Demoting primary");
+            cluster.stepDown();
+
+            // Wait until the cursor address doesn't match the read preference (primary)
+            await().atMost(30, SECONDS)
+                    .pollInterval(1, SECONDS)
+                    .until(() -> cluster.tryPrimary()
+                            .map(node -> !node.getAddress().equals(cursor.getServerAddress()))
+                            .orElse(false));
+
+            // Ensure it's invalid
+            if (!MongoDbNode.IMAGE_VERSION.equals("4.0")) {
+                assertThat(isSelectedReadPreference(client, collection, cursor)).isFalse();
```

Comment:
I'm not sure if this was a client behavior that has since been changed, but it acts differently compared to the later versions.